### PR TITLE
Split function and more numpy-esque slice

### PR DESCRIPTION
**Pull request by jwa7**
**1. Split**

This PR introduces the ``split`` and ``split_block`` functions, which are essentially wrappers for the private function ``_slice_block`` found in operations/slice.py. They allow tensors (TensorMap or TensorBlock, respectively) to be split into multiple tensors according to a specified index name along either the "samples" or "properties" axis, and returned in a list.

How they work:
```py
import numpy as np
from equistore import io, Labels
from equistore.operations import split

# Load a tensor - i.e. SOAP for 11 training structures (idxs 0 -> 11 inc.)
tensor = io.load("path/to/tensor.npz")

split_tensor = split(
    tensor,
    axis="samples",
    grouped_idxs=[
        Labels(
            names=["structure"],
            values=np.array([[0], [6], [7]])
        ),
        Labels(
            names=["structure"],
            values=np.array([[2], [3], [4]])
        ),
        Labels(
            names=["structure"],
            values=np.array([[1], [5], [8], [9], [10]])
        ),
    ]
)
```
In this case, there are 3 groups of structure indices passed as a list of Labels. Inclusion of the ``axis`` argument means that splitting can only occur along one axis with a single function call, in contrast to the slice function where one can specify both the ``samples`` and ``properties`` Labels to slice with. This is intended, so as to mirror the ``join`` function, for which this operation is essentially the inverse.

The ``split*`` functions wrap the private ``_slice_block`` fxn as opposed to the public ``slice*`` fxns as all appropriate input checks are performed in ``split*`` and so would be run twice by calling ``slice*``. To split a tensor into N tensors, N slice operations have to be performed. This is not the ideal under-the-hood implementation, but no better solution could be found at present.

The ``split`` function provides a basis for a ``split_tensors`` function in equisolve that can be used to perform splitting of multiple TensorMaps at once, i.e. for a train-test(-validation) split.

**2. Slice**

Also included in the same PR are changes to slice and appropriate updates to unit tests. The ``numpy.slice`` function does not raise warnings when slicing arrays to emptiness, so after a discussion with @PicoCentauri it was decided that such warnings should be removed from ``equistore.slice`` too. Checks for the arguments to ``slice`` and ``slice_block`` functions were unified, and the ``_slice_tensormap`` function was removed. Unit tests related to error messages were updated with new messages, and those related to warnings were removed.

### CODE REVIEW

File: python/src/equistore/operations/split.py
Comment (DavideTisi) [inline python/src/equistore/operations/split.py:182]:
```suggestion
            "``grouped_idxs`` should be passed as a ``list`` of equistore ``Labels``"
```

File: python/src/equistore/operations/split.py
Comment (DavideTisi) [inline python/src/equistore/operations/split.py:205]:
```suggestion
    for ref_name in ref_names:
```

File: python/tests/operations/split.py
Comment (DavideTisi) [inline python/tests/operations/split.py:520]: this function is very nice, I would put it in some _utilis file in src not hidden here

Comment (DavideTisi) [reply python/tests/operations/split.py:520]: but it should also check the label.names

**Review by DavideTisi — COMMENTED** [review]
General very good just I would move the _label_equal in a more easy to find place. The function is quite general but here it is a bit hidden.

File: python/tests/operations/split.py
Comment (jwa7) [reply python/tests/operations/split.py:521]: Do you mean in ``equistore/python/src/equistore/operations/_utils.py``?

**Review by DavideTisi — APPROVED** [review]
for me it is ok

**Comment by jwa7** [comment]
Thanks for the review @DavideTisi, I think all is good with tests now, right?  If so I'll wait for approval/changes from @PicoCentauri then squash

**Comment by DavideTisi** [comment]
Yes for me everything is ok

Il giorno dom 12 feb 2023 alle 19:02 Joseph W. Abbott <
***@***.***> ha scritto:

File: python/src/equistore/operations/_utils.py
Comment (PicoCentauri) [inline python/src/equistore/operations/_utils.py:133]: How does this function go with @DavideTisi  changes in #121? We maybe can include `_labels_equal ` into #121...

Comment (PicoCentauri) [inline python/src/equistore/operations/_utils.py:133]: How does this function go with @DavideTisi  changes in #121? We maybe can include `_labels_equal ` into #121...

File: python/src/equistore/operations/slice.py
Comment (PicoCentauri) [inline python/src/equistore/operations/slice.py:295]: What is happening if we don't have this check? Are the slice and split function only returning a copy of the original Block? If yes we might think about removing this test.

File: python/src/equistore/operations/split.py
Comment (PicoCentauri) [inline python/src/equistore/operations/split.py:169]: Why is there another check_args function here? The name might be confusing with the one from the slice functions even of they are in different files.

**Review by PicoCentauri — CHANGES_REQUESTED** [review]
I am also almost happy. We should only avoid naming functions equally even if they are in different files. 


Some ideas that go in the spirit of these `_check_args` args functions. So far we only have them here, right?
It might be worth to write a general decorator `@check_args` that checks the arguments of a decorated function based on the type hints. This is a bit of work but could be on the one hand really helpful for the users. On the other it destroys a bit the idea of Python to be vague with input types...

File: python/src/equistore/operations/slice.py
Comment (jwa7) [reply python/src/equistore/operations/slice.py:295]: Currently, yes if we take out this function ``slice(tensor, samples=None, properties=None)`` will return a copy of the input tensor. Is this the behaviour we want? Or do we want to throw an error if neither samples or properties are specified?

File: python/src/equistore/operations/split.py
Comment (jwa7) [reply python/src/equistore/operations/split.py:169]: They are called the same thing but do different things in ``slice.py`` and ``split.py``. I can change the names but don't see much point as they just check arguments in each respective file and should be hidden

File: python/src/equistore/operations/slice.py
Comment (jwa7) [reply python/src/equistore/operations/slice.py:295]: Similarly if I call ``split(tensor, axis="samples", grouped_idxs=[])`` should this also return the original tensor, or throw an error?

@DavideTisi thoughts?

File: python/src/equistore/operations/_utils.py
Comment (jwa7) [reply python/src/equistore/operations/_utils.py]: I don't think @DavideTisi included it in #121, which has now been merged. Shall I keep it in ``_utils.py`` for this PR?

**Review by PicoCentauri — APPROVED** [review]
Thanks for the changes @jwa7!

**Comment by jwa7** [comment]
Thanks for the reviews @DavideTisi @PicoCentauri !

**Pull request merged**
